DM-27877: Write technote summarizing crowded field analysis (#1)
lsst-dm/dmtn-171: index.rst

Crucially, the AP processing must meet its requirements even in the most crowded fields it will encounter.
Of particular importance are the transient reporting latency (OTT1 and OTR1; LSR-REQ-0117 and LSR-REQ-0025), 
the reliability with which images are processed (sciVisitAlertDelay and sciVisitAlertFailure; OSS-REQ-0112), 
the photometric and astrometric accuracy of the resulting catalogs (dmL1AstroErr, dmL1PhotoErr, and photoZeroPointOffset; OSS-REQ-0149 and OSS-REQ-0152), 
completeness and purity metrics for transients (OSS-REQ-0353) and solar system objects (OSS-REQ-0354),
and source misassociation for transients (OSS-REQ-0160) and solar system objects (OSS-REQ-0159).
Accordingly, large-scale tests on precursor datasets are necessary to identify where improvements are needed in order to meet requirements.
The status of the processing also will inform ongoing cadence planning for the overall survey.

The scope of this technote does not include obtaining photometry for static sources in direct images (except insofar as it is required to conduct image differencing), nor comparisons to dedicated crowded field processing codes.

> `DMTN-077 <https://dmtn-077.lsst.io>`_ addressess those issues.

Reviewer: ```suggestion
`DMTN-077 <https://dmtn-077.lsst.io>`_ addresses those issues.
```

Some of the earliest applications of image differencing in astronomy were motivated by the desire to identify transient sources (microlensing events) in extremely crowded stellar fields.
At ground-based photometric precision, typically about one percent of stellar sources have detectable variability. 
Thus by subtracting a coadded historical (template) image from a new (science) image, the crowding can be reduced by a factor of ~100.

In this technote we assess the state of crowded field processing in Fall 2020 using the LSST Alert Production (AP) Science Pipelines.
Alert Production uses image differencing to identify transient, moving, and variable sources in LSST images within 60 seconds of camera readout.
Crucially, the AP processing must meet its requirements even in the most crowded fields it will encounter.
Of particular importance are the transient reporting latency (OTT1 and OTR1; LSR-REQ-0117 and LSR-REQ-0025), 
the reliability with which images are processed (sciVisitAlertDelay and sciVisitAlertFailure; OSS-REQ-0112), 
the photometric and astrometric accuracy of the resulting catalogs (dmL1AstroErr, dmL1PhotoErr, and photoZeroPointOffset; OSS-REQ-0149 and OSS-REQ-0152), 

> completeness and purity metrics for transients (OSS-REQ-0353) and solar system objects ( OSS-REQ-0354),

Reviewer: ```suggestion
completeness and purity metrics for transients (OSS-REQ-0353) and solar system objects (OSS-REQ-0354),
```

and source misassociation for transients (OSS-REQ-0160) and solar system objects (OSS-REQ-0159).
Accordingly, large-scale tests on precursor datasets are necessary to identify where improvements are needed in order to meet requirements.
The status of the processing also will inform ongoing cadence planning for the overall survey.

The scope of this technote does not include obtaining photometry for static sources in direct images (except insofar as it is required to conduct image differencing), nor comparisons to dedicated crowded field processing codes.

> `DMTN-077 <https://dmtn-077.lsst.io>`_ addressess these issues.

Reviewer: ```suggestion
`DMTN-077 <https://dmtn-077.lsst.io>`_ addressess those issues.
```

Overview of the data
====================

> To test the LSST Science Pipelines on crowded fields, we acquired public data from the DECam survey of the galactic bulge by `Saha et. al <https://arxiv.org/pdf/1902.05637.pdf>`_.

Reviewer: ```suggestion
To test the LSST Science Pipelines on crowded fields, we acquired public data from the DECam survey of the galactic bulge by `Saha et. al 2019<https://arxiv.org/pdf/1902.05637.pdf>`_.
```

The full survey consists of six fields at low galactic latitudes observed in the six ``ugrizy`` bands, though for this investigation we limited the analysis to g- and i-bands for the two fields closest to the galactic center (B1 and B2, :numref:`field_coordinates`).
The two fields selected for this test were observed in 2013 and 2015 to similar depth in both bands (:numref:`visit_summary`).
These data are available on ``lsst-devl`` at ``/datasets/decam``, while the processed data and the scripts that produced them available at ``/project/sullivan/saha2``.

> The size of the processed data amounts to 7.0Tb on disk, which can be fully recreated following the instructions in ``/project/sullivan/saha2/README``.

Reviewer: Could we put the README somewhere more permanent/accessible?  Maybe even in the repository with this technote?

Reviewer: Also I would explicitly state which version(s) of the stack you used for the processing, whether here or somewhere else in the technote.

Author: I've added a reference to the software version I used (release 21.0), but I am unsure how to best include information from the README. I can see how someone reading this technote might be curious and want to read it, but not have access to lsst-devl, but the README itself doesn't make much sense out of context since it refers to directories and scripts in that location. For now, I've tweaked the wording to make it more clear that the README gives instructions on how to run the scripts.

Reviewer 2: I would suggest that all the `.conf`, `.sh`, and `.sl` files are worthy of being preserved in GitHub, either in the DMTN repo or a separate `lsst-dm` repo.  The few references to `lsst-devl`-specific locations (primarily `/software/lsstsw`, `/datasets/decam`, and `/project/sullivan/saha2` itself, I believe) can be marked or an environment variable substituted.  References to `/project/sullivan/code/ip_isr` (if actually needed) should instead refer to a particular release or at least GitHub hash.

.. table:: Summary of visits
   :name: visit_summary
    
   ====== ====== ====== ======
   Field  band   2013   2015
   ====== ====== ====== ======
   B1       g    30     44
   B2       g    31     44
   B1       i    53     44
   B2       i    49     44
   ====== ====== ====== ======

It is important to note that the seeing was uniformly worse in 2013 than in 2015 for this survey (:numref:`seeing_B1_g` - :numref:`seeing_B2_i`).
As a result, we expect image differencing to fare much worse when running ``ap_pipe`` on the 2015 observations using the template constructed from the 2013 observations.

.. figure:: /_static/Seeing_B1_g.png
 :name: seeing_B1_g

 Distribution of seeing for g-band observations of field B1

.. figure:: /_static/Seeing_B2_g.png
 :name: seeing_B2_g

 Distribution of seeing for g-band observations of field B2

.. figure:: /_static/Seeing_B1_i.png
 :name: seeing_B1_i

 Distribution of seeing for i-band observations of field B1

.. figure:: /_static/Seeing_B2_i.png
 :name: seeing_B2_i

 Distribution of seeing for i-band observations of field B2

Running the Science Pipelines
=============================

> Running the LSST Science Pipelines on observations of crowded fields follows the same steps as for fields with lower densities, but a few of those steps require modification to work properly.

Reviewer: ```suggestion
Running the LSST Science Pipelines on observations of crowded fields follows the same steps as for fields with lower densities, but in this investigation a few of those steps required modification to work properly.
```

Here, we divide processing into three distinct stages:

1. **Single Frame Processing**: includes applying flat field and bias corrections, Instrument Signature Removal, and astrometric and photometric calibration.
2. **Coaddition**: includes warping and PSF-matching the individual exposures, as well as the actual coaddition algorithm.
3. **Alert Production**: includes all of Single Frame Processing for the science images, as well as warping the templates, performing image differencing, detecting and measuring sources, and associating sources from different visits to form objects.

Single Frame Processing
-----------------------

> The initial stages of Single Frame Processing require no modifications to accomodate crowded fields.

Reviewer: ```suggestion
The initial stages of Single Frame Processing required no modifications to accommodate crowded fields.
```

> Once processing moves on to the first instance of source detection, however, it becomes important to set the detection threshold to a fixed value rather than the default of a multiple of the standard deviation of the background.

Reviewer: ```suggestion
In source detection, however, it became necessary to set the detection threshold to a fixed value rather than the default of a multiple of the standard deviation of the background.
```

There may be no pixels without a detectable source in the exposures, so the measured background level will be incorrect and the number of sources used for PSF modeling will be unpredictable, and possibly too few.
For this test, we took typical detection thresholds from DECam HiTS observations and found that those eliminated the related processing errors.
Further refinement would likely yield improved results.

> All of the modifications needed to run single frame processing on these data can be found in :numref:`processCcd_config`, below.

Reviewer: I'd say explicitly here that these are configuration changes relative to a specific version of the stack.

.. table:: Modified config settings needed for single frame processing
   :name: processCcd_config

   ============================================== ======== ======
   Modified config settings for processCcd.py     value    band 
   ============================================== ======== ======
   charImage.requireCrForPsf                      False    i, g
   charImage.detection.thresholdValue             10000    i    
   charImage.detection.thresholdValue             2500     g    
   charImage.detection.includeThresholdMultiplier 1.0      i, g  
   charImage.detection.thresholdType              "value"  i, g  
   charImage.repair.cosmicray.nCrPixelMax         10000000 i, g  
   charImage.repair.cosmicray.min_DN              10000    i    
   charImage.repair.cosmicray.min_DN              2500     g    
   ============================================== ======== ======

> Beyond the source detection thresholds, it was necessary to modify two additional components.

Reviewer: I would give this paragraph a slightly different spin: PSFEx was already supposed to be the default (per RFC-312), it just wasn't implemented yet for all cameras.  We didn't (and wouldn't have) changed the default PSF estimator for all cameras just because of this crowded field study.

I'd simply say that for the version of the stack we were using for the processing, PSFEx hadn't been set as the default, and we fixed that.

We found that the default algorithm for measuring the PSF, a simple PCA-based model, simply failed when run on most of the visits from these crowded fields.
However, PSFex was able to successfully measure the PSF, and since it was already available in the Science Pipelines we made it the default for all cameras.
Thus, no further modifications are needed for future processing.

> The final component that requires modification is the cosmic ray detection and repair algorithm.

Reviewer: ```suggestion
The final component that required modification is the cosmic ray detection and repair algorithm.
```

In these cases the failure appears to be due to every pixel in the image being identified as a cosmic ray.
This failure suggests that our cosmic ray detection algorithm needs improvement and should be investigated further, but because of the low number of exposures affected we simply increased the number of pixels required to trigger the failure.
This does not solve the problem, but it allows us to continue processing these exposures to make sure that there are no additional problems.

Evaluation of the Point Spread Function (PSF)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The accuracy of the measurement of the Point Spread Function (PSF) is our greatest concern with processing crowded fields, since it is typically not possible to find a sufficient number of isolated stellar sources to measure.
The PSF is used for very little in the current Science Pipelines; our standard Alard&Lupton-style image differencing depends only on the calculated size of the PSF to compare with that of the template, and not on the shape of the PSF.

> However, the accuracy of the PSF does impact source measurement and many science use cases.

Reviewer: ```suggestion
However, the accuracy of the PSF does impact source measurement.
```

Reviewer: (I'm suggesting we drop "science use cases" here because the AP catalog products are not made available in any systematic way to users)

> It is likely that the current implementation of PSFex is sufficient for internal processing of crowded field data, but scientists with strict requirements on the quality of the PSF would be advised to measure the PSF independently in post-processing with a carefully tuned algorithm.

Reviewer: ```suggestion
```

Reviewer: This feels speculative to me--let's just strike it for now.

.. figure:: /_static/psf_B1_2013_i.png
 :name: psf_B1_2013_i

 PSFs for each of the i-band visits from 2013 in field B1, for a CCD in the center of the focal plane.

.. figure:: /_static/psf_B2_2013_i.png
 :name: psf_B2_2013_i

 PSFs for each of the i-band visits from 2013 in field B2, for a CCD in the center of the focal plane.

.. figure:: /_static/psf_B1_2015_g.png
 :name: psf_B1_2015_g

 PSFs for each of the g-band visits from 2015 in field B1, for a CCD in the center of the focal plane.

.. figure:: /_static/psf_B2_2015_g.png
 :name: psf_B2_2015_g

 PSFs for each of the g-band visits from 2015 in field B2, for a CCD in the center of the focal plane.

.. figure:: /_static/psf_B1_2015_i.png
 :name: psf_B1_2015_i

 PSFs for each of the i-band visits from 2015 in field B1, for a CCD in the center of the focal plane.

.. figure:: /_static/psf_B2_2015_i.png
 :name: psf_B2_2015_i

 PSFs for each of the i-band visits from 2015 in field B2, for a CCD in the center of the focal plane.

While g-band generally has clean and reasonably symmetric-looking PSFs, some i-band visits show worrisome features in the wings of the PSF.

> As noted above, these are not likely to impact the performance of the Science Pipelines, though it is undesirable and will likely impact downstream science users.

Reviewer: ```suggestion
As noted above, these are not likely to impact the performance of Alert Production, though it is undesirable.
```

> For these crowded fields, our current PSF modeling algorithm PSFex is sufficient to run the Science Pipelines, but a more sophisticated algorithm would be desireable.

Reviewer: ```suggestion
For these crowded fields, our current PSF modeling algorithm PSFex is sufficient to run Alert Production, but a more robust algorithm would be desirable.
```

Density of measured sources on a single ccd
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Perhaps the most important metric for evaluating the performance of the Science Pipelines on crowded fields is the measured density of sources.

Reviewer: I'm not actually sure if this section is necessary for this technote (though it's interesting)--we don't necessarily need SFM to detect a complete sample of sources, just enough that we can calibrate correctly prior to diffim.  With that said these plots are good diagnostics for calibration errors, as you point out.   I'd suggest tuning this paragraph slightly to that end.

Author: I've changed the wording to try and set the context better. Most of interesting information is in the next `Source Counts` section, though, so I could just drop this section.

   +------+------+-------+------------------------------------------------+-----------------------------------+
   | 2013 | i    | B2    | 210559, 210601, 210637, 210673, 216048         | :numref:`source_counts_2013_B2_i` |
   +------+------+-------+------------------------------------------------+-----------------------------------+
   | 2015 | i    | B2    | 427624                                         | :numref:`source_counts_2015_B2_i` |
   +------+------+-------+------------------------------------------------+-----------------------------------+

Timing
^^^^^^

A final concern is the amount of time it will take to process each ccd in crowded fields.
While a typical ccd took just under 4 minutes to process, there was a long tail of ccds that took far longer (:numref:`Timing_2013` and :numref:`Timing_2015`).
The increased time was entirely spent in two steps: matching the detected objects to a reference catalog, and measuring the difference image sources.
The time required for matching appeared to be non-linear, with the ccds with the largest number of sources and reference objects to match requiring up to four hours to complete.

> Our matching algorithm was not designed for these very large numbers of sources, so we are encouraged by the results even if the performance is slow. 

Reviewer: Chris definitely tested the matcher on the New Horizons Bulge data, so I do think it was designed for dense fields.  I'd simply say that  further performance tuning is needed.

Author: The phrasing here was based on a conversation with Chris where he was surprised by the very high density of sources I was getting. I will change the wording to reflect that it was not optimized for these densities, instead of that it was not designed for them.

 The color scale is locked to the scale of the template in :numref:`Template_B2_2013_i`

.. figure:: /_static/Compare_216144_i_42_template.png
 :name: Template_B2_2013_i

 Deep coadd template for i-band visit 216144 ccd 42.
 The color scale uses a Asinh stretch to emphasize faint features.

.. figure:: /_static/Compare_216144_i_42_diff.png
 :name: Diffim_B2_2013_i

 Image difference for i-band visit 216144 ccd 42.
 The color scale is locked to the scale of the template in :numref:`Template_B2_2013_i`

.. figure:: /_static/Compare_223465_g_42_calexp.png
 :name: Calexp_B2_2013_g

 G-band science visit 223465 ccd 42 from 2013 B2.
 The color scale is locked to the scale of the template in :numref:`Template_B2_2013_g`

.. figure:: /_static/Compare_223465_g_42_template.png
 :name: Template_B2_2013_g

 Deep coadd template for g-band visit 223465 ccd 42.
 The color scale uses a Asinh stretch to emphasize faint features.

.. figure:: /_static/Compare_223465_g_42_diff.png
 :name: Diffim_B2_2013_g

 Image difference for g-band visit 223465 ccd 42.
 The color scale is locked to the scale of the template in :numref:`Template_B2_2013_g`

Several features are apparent from the above images.

> Most importantly, despite the sea of overlapping sources in the input images and the imperfect subtraction, the residuals are still isolated.

Reviewer: ```suggestion
Most importantly, despite the sea of overlapping sources in the input images and the imperfect subtraction, the resulting DIASources are still isolated.
```

Thus, we can still detect and measure sources in the difference image, though we have far more to deal with than for a typical observation.

> Most of the residuals are also unfortunately typical of those we see in less crowded fields, which is indicative of our need to improve the implementation our image differencing algorithm in general.

Reviewer: ```suggestion
Most of the DIASources show artifacts characteristic of imperfect subtraction, such as dipoles and ringing patterns.
These features are also unfortunately common in current processing of less crowded fields, which is indicative of our need to improve the implementation our image differencing in general.
```

   :name: dia_source_density_B2_i
  
   Density of detected DIA sources across all visits for field B2 in i-band, for ccd 42.
   Compare to :numref:`source_density_B2_i` for the number of sources in the original images.

Conclusions and future work
===========================

This investigation has stress-tested the LSST Science Pipelines, and uncovered several algorithmic components that need attention.
Some of those improvements, such as upgrading the PSF determiner, were necessary to process the data and have already been completed.
Others, such as the fidelity of image differencing, had been previously identified and the improvements are under active development.

Summary of the challenges to processing crowded fields identified in this analysis:

- The PSF determiner was upgraded to PSFex, which is able to run on crowded fields. However, it does not appear to be able to model the wings of the PSF (see :numref:`psf_B1_2013_g` through :numref:`psf_B2_2015_i`).
- The cosmic ray detection and repair algorithm still fails for some ccds, and will require either careful tuning of the existing parameters or a more sophisticated implementation.

> - Photometric calibration is at times inconsistent, with offsets of several magnitudes in the worst cases (:numref:`source_counts_2013_B1_g` - :numref:`source_counts_2015_B2_i`). This is likely to improve with a better PSF model.

Reviewer: It's not clear to me that PSF modeling would cause the problems with photometric calibration--let's not speculate here.

> - The improved PSF model will also be needed to model bright stars.

Reviewer: Why do we need to model bright stars for AP?

Author: I was thinking of background subtraction in the vicinity of bright stars, but that is not a subject I covered in this technote. I'll just remove this line.

- We are able to measure sources at densities greater than 500,000 per square degree under good conditions, and the source counts suggest that we are not undercounting sources.

> - Future work should inject fake sources in the analysis to measure completeness.

Reviewer: ```suggestion
- In future processing of these data we will use newly-developed pipelines to inject fake sources in the analysis to measure completeness, and we will perform a systematic visual inspection to determine purity.
```
